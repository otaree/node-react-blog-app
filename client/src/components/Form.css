.form {
    width: 80%;
    margin: 10px 15%;
}

.form-control .form-input, 
.form-control .form-textarea
 {
    outline: none;
    width: 80%;
    margin: 10px 0;
    border: 1px solid #888;
    border-radius: 3px;
    height: 25px;
    padding-left: 10px;
}

.form-control .form-textarea {
    padding-top: 10px;
    height: 300px;
}

.form-control .form-input:focus, .form-control .form-textarea:focus {
    border: 2px solid rgb(102, 156, 173);    
}

.form-select label {
    font-size: 16px;
    font-weight: bold;
    margin-right: 10px;
}


.form-select select {
    outline: none;
    border: thin solid #888;
    border-radius: 3px;
    width: 70px;
    padding-left: 10px;
    box-sizing: border-box;
    font-size: 14px;
    background-color: #ccc;
    margin-bottom: 10px;
}

.submit {
    margin: 0 35%;
}

.btn-submit {
    outline: none;
    border: 1px solid #888;
    font-size: 16px;
    background-color: rgb(55, 148, 55);
    color: #eee;
    border-radius: 3px;
    font-weight: bold;
    padding: 10px;
    width: 90px;
}

.btn-submit:hover {
    background-color: lightgreen;
    color: #444;
    cursor: pointer;
}