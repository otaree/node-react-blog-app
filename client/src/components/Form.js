import React from 'react';
import './Form.css';

export default class Form extends React.Component {
    state = {
        title: {
            value: this.props.blog ? this.props.blog.title : "",
            error: false,
        },
        author: {
            value: this.props.blog ? this.props.blog.author : "",
            error: false,
        },
        body: {
            value: this.props.blog ? this.props.blog.body : "",
            error: false,
        },
        published: this.props.blog ? this.props.blog.published : false       
    };

    onTitleChange = e => {
        const value = e.target.value;
        this.setState(prevState => {
            return {
                title: {
                    ...prevState.title,
                    value
                }
            };
        });
    };

    onBodyChange = e => {
        const value = e.target.value;
        this.setState(prevState => {
            return {
                body: {
                    ...prevState.body,
                    value
                }
            };
        });
    };

    onAuthorChange = e => {
        const value = e.target.value;
        this.setState(prevState => {
            return {
                author: {
                    ...prevState.author,
                    value
                }
            };
        });
    };

    onPublishChange = e => {
        const value = e.target.value;
        this.setState(prevState => {
            return {
                    published: value,
                }
            });
    };

    onSubmit = (e) => {
        e.preventDefault();

        let isError = false;

        if (this.state.title.value.trim() === '') {
            isError = true;
            this.setState(prevState => {
                return {
                    title: {
                        ...prevState.title,
                        error: true,
                    }
                }
            });
        }


        if (this.state.body.value.trim() === '') {
            isError = true;
            this.setState(prevState => {
                return {
                    body: {
                        ...prevState.body,
                        error: true,
                    }
                }
            });
        }

        if (this.state.author.value.trim() === '') {
            isError = true;
            this.setState(prevState => {
                return {
                    author: {
                        ...prevState.author,
                        error: true,
                    }
                }
            });
        }

        if (isError) return;
        this.setState(prevState => {
            return {
                title: {
                    ...prevState.title,
                    error: false,
                },
                body: {
                    ...prevState.body,
                    error: false,
                },
                author: {
                    ...prevState.author,
                    error: false,
                },
            };
        });
        this.props.onSubmit({
            title: this.state.title.value,
            body: this.state.body.value,
            author: this.state.author.value,
            published: this.state.published
        });
    }

    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                <div className="form-control">
                    <input 
                        className="form-input"
                        type="text"
                        placeholder="Title"
                        value={this.state.title.value}
                        onChange={this.onTitleChange}
                    />
                    { this.state.title.error && <div>Invalid title</div>}
                </div>
                <div className="form-control">
                    <textarea 
                        className="form-textarea"                    
                        placeholder="Body"
                        value={this.state.body.value}
                        onChange={this.onBodyChange}
                    />
                    { this.state.body.error && <div>Invalid body</div> }
                </div>
                <div className="form-control">
                    <input 
                        className="form-input"                        
                        type="text"
                        placeholder="Author"
                        value={this.state.author.value}
                        onChange={this.onAuthorChange}
                    />
                    { this.state.author.error && <div>Invalid author name</div> }
                </div>
                <div className="form-select">
                    <label>publish</label>
                    <select name="publish">
                        <option value={this.state.published} onChange={this.onPublishChange}>false</option>
                        <option value="true">True</option>
                    </select>
                </div>
                <div className="submit">
                    <button className="btn-submit">Submit</button>
                </div>
            </form>
        );
    }
}