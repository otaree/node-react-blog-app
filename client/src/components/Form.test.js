import React from 'react';
import { shallow } from 'enzyme';
import Form  from './Form';

describe('Form', () => {
    let form = shallow(<Form />);
    it('renders properly', () => {
        expect(form).toMatchSnapshot();
    });

    describe('when form field value changes', () => {
        beforeEach(() => {
            form = shallow(<Form />);
        });

        it('should set title on input change', () => {
            const value = "New Title";
            form.find('input').at(0).simulate('change', {
                target: { value }
            });
            expect(form.state('title').value).toEqual(value);
        });

        it('should set body on input change', () => {
            const value = "New Body";
            form.find('textarea').simulate('change', {
                target: { value }
            });
            expect(form.state('body').value).toEqual(value);
        });

        it('should set author on input change', () => {
            const value = "Author";
            form.find('input').at(1).simulate('change', {
                target: { value }
            });
            expect(form.state('author').value).toEqual(value);
        })
    });

    describe('when invalid form submission', () => {
        beforeEach(() => {
            form = shallow(<Form />);
        });

        it('should render error', () => {
            form.find('form').simulate('submit', {
                preventDefault: () => {}
            });
            expect(form).toMatchSnapshot();
        });

        it('should render set error true for all fields', () => {
            form.find('form').simulate('submit', {
                preventDefault: () => {}
            });
            expect(form.state('title').error).toEqual(true);
            expect(form.state('body').error).toEqual(true);
            expect(form.state('author').error).toEqual(true);
        });

        it('should render error for all expect title', () => {
            const value = 'New Title';
            form.find('input').at(0).simulate('change', {
                target: { value }
            });
            form.find('form').simulate('submit', {
                preventDefault: () => {}
            });
            expect(form.state('title').error).toEqual(false);
            expect(form.state('body').error).toEqual(true);
            expect(form.state('author').error).toEqual(true);
        })
    });

    describe('when valid form submission', () => {
        const mockOnSubmit = jest.fn();        
        beforeEach(() => {
            const props = {
                onSubmit: mockOnSubmit
            }
            form = shallow(<Form {...props} />);
        });
        
        it('should call props `onSubmit`', () => {
            const title = "Title";
            const body = "body";
            const author = "Author";
            form.find('input').at(0).simulate('change', {
                target: {
                    value: title
                }
            });
            form.find('textarea').simulate('change', {
                target: {
                    value: body
                }
            });
            form.find('input').at(1).simulate('change', {
                target: {
                    value: author
                }
            });
            form.find('form').simulate('submit', {
                preventDefault: () => {}
            });
            expect(mockOnSubmit).toHaveBeenCalledWith({
                title,
                body,
                author,
                published: false
            });
        });
    });
});