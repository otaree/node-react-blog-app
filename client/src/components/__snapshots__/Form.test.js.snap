// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="form-control"
>
          <input
                    className="form-input"
                    onChange={[Function]}
                    placeholder="Title"
                    type="text"
                    value=""
          />
</div>,
        <div
          className="form-control"
>
          <textarea
                    className="form-textarea"
                    onChange={[Function]}
                    placeholder="Body"
                    value=""
          />
</div>,
        <div
          className="form-control"
>
          <input
                    className="form-input"
                    onChange={[Function]}
                    placeholder="Author"
                    type="text"
                    value=""
          />
</div>,
        <div
          className="form-select"
>
          <label>
                    publish
          </label>
          <select
                    name="publish"
          >
                    <option
                              onChange={[Function]}
                              value={false}
                    >
                              false
                    </option>
                    <option
                              value="true"
                    >
                              True
                    </option>
          </select>
</div>,
        <div
          className="submit"
>
          <button
                    className="btn-submit"
          >
                    Submit
          </button>
</div>,
      ],
      "className": "form",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="form-input"
              onChange={[Function]}
              placeholder="Title"
              type="text"
              value=""
/>,
            false,
          ],
          "className": "form-control",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-input",
              "onChange": [Function],
              "placeholder": "Title",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          false,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <textarea
              className="form-textarea"
              onChange={[Function]}
              placeholder="Body"
              value=""
/>,
            false,
          ],
          "className": "form-control",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-textarea",
              "onChange": [Function],
              "placeholder": "Body",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "textarea",
          },
          false,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="form-input"
              onChange={[Function]}
              placeholder="Author"
              type="text"
              value=""
/>,
            false,
          ],
          "className": "form-control",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-input",
              "onChange": [Function],
              "placeholder": "Author",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          false,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label>
              publish
</label>,
            <select
              name="publish"
>
              <option
                            onChange={[Function]}
                            value={false}
              >
                            false
              </option>
              <option
                            value="true"
              >
                            True
              </option>
</select>,
          ],
          "className": "form-select",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "publish",
            },
            "ref": null,
            "rendered": "publish",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option
                  onChange={[Function]}
                  value={false}
>
                  false
</option>,
                <option
                  value="true"
>
                  True
</option>,
              ],
              "name": "publish",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "false",
                  "onChange": [Function],
                  "value": false,
                },
                "ref": null,
                "rendered": "false",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "True",
                  "value": "true",
                },
                "ref": null,
                "rendered": "True",
                "type": "option",
              },
            ],
            "type": "select",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="btn-submit"
>
            Submit
</button>,
          "className": "submit",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Submit",
            "className": "btn-submit",
          },
          "ref": null,
          "rendered": "Submit",
          "type": "button",
        },
        "type": "div",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="form-control"
>
            <input
                        className="form-input"
                        onChange={[Function]}
                        placeholder="Title"
                        type="text"
                        value=""
            />
</div>,
          <div
            className="form-control"
>
            <textarea
                        className="form-textarea"
                        onChange={[Function]}
                        placeholder="Body"
                        value=""
            />
</div>,
          <div
            className="form-control"
>
            <input
                        className="form-input"
                        onChange={[Function]}
                        placeholder="Author"
                        type="text"
                        value=""
            />
</div>,
          <div
            className="form-select"
>
            <label>
                        publish
            </label>
            <select
                        name="publish"
            >
                        <option
                                    onChange={[Function]}
                                    value={false}
                        >
                                    false
                        </option>
                        <option
                                    value="true"
                        >
                                    True
                        </option>
            </select>
</div>,
          <div
            className="submit"
>
            <button
                        className="btn-submit"
            >
                        Submit
            </button>
</div>,
        ],
        "className": "form",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="form-input"
                onChange={[Function]}
                placeholder="Title"
                type="text"
                value=""
/>,
              false,
            ],
            "className": "form-control",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-input",
                "onChange": [Function],
                "placeholder": "Title",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            false,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <textarea
                className="form-textarea"
                onChange={[Function]}
                placeholder="Body"
                value=""
/>,
              false,
            ],
            "className": "form-control",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-textarea",
                "onChange": [Function],
                "placeholder": "Body",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "textarea",
            },
            false,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="form-input"
                onChange={[Function]}
                placeholder="Author"
                type="text"
                value=""
/>,
              false,
            ],
            "className": "form-control",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-input",
                "onChange": [Function],
                "placeholder": "Author",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            false,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label>
                publish
</label>,
              <select
                name="publish"
>
                <option
                                onChange={[Function]}
                                value={false}
                >
                                false
                </option>
                <option
                                value="true"
                >
                                True
                </option>
</select>,
            ],
            "className": "form-select",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "publish",
              },
              "ref": null,
              "rendered": "publish",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <option
                    onChange={[Function]}
                    value={false}
>
                    false
</option>,
                  <option
                    value="true"
>
                    True
</option>,
                ],
                "name": "publish",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "false",
                    "onChange": [Function],
                    "value": false,
                  },
                  "ref": null,
                  "rendered": "false",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "True",
                    "value": "true",
                  },
                  "ref": null,
                  "rendered": "True",
                  "type": "option",
                },
              ],
              "type": "select",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="btn-submit"
>
              Submit
</button>,
            "className": "submit",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "className": "btn-submit",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
          "type": "div",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`Form when invalid form submission should render error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="form-control"
>
          <input
                    className="form-input"
                    onChange={[Function]}
                    placeholder="Title"
                    type="text"
                    value=""
          />
          <div>
                    Invalid title
          </div>
</div>,
        <div
          className="form-control"
>
          <textarea
                    className="form-textarea"
                    onChange={[Function]}
                    placeholder="Body"
                    value=""
          />
          <div>
                    Invalid body
          </div>
</div>,
        <div
          className="form-control"
>
          <input
                    className="form-input"
                    onChange={[Function]}
                    placeholder="Author"
                    type="text"
                    value=""
          />
          <div>
                    Invalid author name
          </div>
</div>,
        <div
          className="form-select"
>
          <label>
                    publish
          </label>
          <select
                    name="publish"
          >
                    <option
                              onChange={[Function]}
                              value={false}
                    >
                              false
                    </option>
                    <option
                              value="true"
                    >
                              True
                    </option>
          </select>
</div>,
        <div
          className="submit"
>
          <button
                    className="btn-submit"
          >
                    Submit
          </button>
</div>,
      ],
      "className": "form",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="form-input"
              onChange={[Function]}
              placeholder="Title"
              type="text"
              value=""
/>,
            <div>
              Invalid title
</div>,
          ],
          "className": "form-control",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-input",
              "onChange": [Function],
              "placeholder": "Title",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Invalid title",
            },
            "ref": null,
            "rendered": "Invalid title",
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <textarea
              className="form-textarea"
              onChange={[Function]}
              placeholder="Body"
              value=""
/>,
            <div>
              Invalid body
</div>,
          ],
          "className": "form-control",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-textarea",
              "onChange": [Function],
              "placeholder": "Body",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "textarea",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Invalid body",
            },
            "ref": null,
            "rendered": "Invalid body",
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="form-input"
              onChange={[Function]}
              placeholder="Author"
              type="text"
              value=""
/>,
            <div>
              Invalid author name
</div>,
          ],
          "className": "form-control",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-input",
              "onChange": [Function],
              "placeholder": "Author",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Invalid author name",
            },
            "ref": null,
            "rendered": "Invalid author name",
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label>
              publish
</label>,
            <select
              name="publish"
>
              <option
                            onChange={[Function]}
                            value={false}
              >
                            false
              </option>
              <option
                            value="true"
              >
                            True
              </option>
</select>,
          ],
          "className": "form-select",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "publish",
            },
            "ref": null,
            "rendered": "publish",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option
                  onChange={[Function]}
                  value={false}
>
                  false
</option>,
                <option
                  value="true"
>
                  True
</option>,
              ],
              "name": "publish",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "false",
                  "onChange": [Function],
                  "value": false,
                },
                "ref": null,
                "rendered": "false",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "True",
                  "value": "true",
                },
                "ref": null,
                "rendered": "True",
                "type": "option",
              },
            ],
            "type": "select",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="btn-submit"
>
            Submit
</button>,
          "className": "submit",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Submit",
            "className": "btn-submit",
          },
          "ref": null,
          "rendered": "Submit",
          "type": "button",
        },
        "type": "div",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="form-control"
>
            <input
                        className="form-input"
                        onChange={[Function]}
                        placeholder="Title"
                        type="text"
                        value=""
            />
            <div>
                        Invalid title
            </div>
</div>,
          <div
            className="form-control"
>
            <textarea
                        className="form-textarea"
                        onChange={[Function]}
                        placeholder="Body"
                        value=""
            />
            <div>
                        Invalid body
            </div>
</div>,
          <div
            className="form-control"
>
            <input
                        className="form-input"
                        onChange={[Function]}
                        placeholder="Author"
                        type="text"
                        value=""
            />
            <div>
                        Invalid author name
            </div>
</div>,
          <div
            className="form-select"
>
            <label>
                        publish
            </label>
            <select
                        name="publish"
            >
                        <option
                                    onChange={[Function]}
                                    value={false}
                        >
                                    false
                        </option>
                        <option
                                    value="true"
                        >
                                    True
                        </option>
            </select>
</div>,
          <div
            className="submit"
>
            <button
                        className="btn-submit"
            >
                        Submit
            </button>
</div>,
        ],
        "className": "form",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="form-input"
                onChange={[Function]}
                placeholder="Title"
                type="text"
                value=""
/>,
              <div>
                Invalid title
</div>,
            ],
            "className": "form-control",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-input",
                "onChange": [Function],
                "placeholder": "Title",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Invalid title",
              },
              "ref": null,
              "rendered": "Invalid title",
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <textarea
                className="form-textarea"
                onChange={[Function]}
                placeholder="Body"
                value=""
/>,
              <div>
                Invalid body
</div>,
            ],
            "className": "form-control",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-textarea",
                "onChange": [Function],
                "placeholder": "Body",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "textarea",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Invalid body",
              },
              "ref": null,
              "rendered": "Invalid body",
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="form-input"
                onChange={[Function]}
                placeholder="Author"
                type="text"
                value=""
/>,
              <div>
                Invalid author name
</div>,
            ],
            "className": "form-control",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-input",
                "onChange": [Function],
                "placeholder": "Author",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Invalid author name",
              },
              "ref": null,
              "rendered": "Invalid author name",
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label>
                publish
</label>,
              <select
                name="publish"
>
                <option
                                onChange={[Function]}
                                value={false}
                >
                                false
                </option>
                <option
                                value="true"
                >
                                True
                </option>
</select>,
            ],
            "className": "form-select",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "publish",
              },
              "ref": null,
              "rendered": "publish",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <option
                    onChange={[Function]}
                    value={false}
>
                    false
</option>,
                  <option
                    value="true"
>
                    True
</option>,
                ],
                "name": "publish",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "false",
                    "onChange": [Function],
                    "value": false,
                  },
                  "ref": null,
                  "rendered": "false",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "True",
                    "value": "true",
                  },
                  "ref": null,
                  "rendered": "True",
                  "type": "option",
                },
              ],
              "type": "select",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="btn-submit"
>
              Submit
</button>,
            "className": "submit",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "className": "btn-submit",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
          "type": "div",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
